### ISSUE TITLE
Segmentation fault when a Tensor backward hook removes itself

### ISSUE THREAD

**Issue by nelhage** [pytorch/pytorch#58354]
## 🐛 Bug

If a Tensor backward hook removes itself during execution *and* returns a modified gradient, PyTorch will segfault.

## To Reproduce

```python
import torch

def install_hook(tensor):
    handle = None
    def hook(tensor):
        handle.remove()
        return torch.zeros_like(tensor)
    handle = tensor.register_hook(hook)

def test_hook():
    t = torch.ones((1, 5))
    t.requires_grad_()
    install_hook(t)
    (t ** 2).mean().backward()
    print(t.grad)

if __name__ == '__main__':
    test_hook()
```

```
$ python test_hook.py 
Segmentation fault (core dumped)
```

Stacktrace, as recorded by gdb:

```
#0  func_get_name (op=0x7fffb186b790, _unused_ignored=0x0) at /tmp/build/80754af9/python_1599203911753/work/Objects/funcobject.c:299
#1  0x0000555555692f65 in _PyObject_GenericGetAttrWithDict (obj=0x7fffb186b790, name=0x7ffff74b1770, dict=0x0, suppress=0) at /tmp/build/80754af9/python_1599203911753/work/Objects/object.c:1254
#2  0x00005555556a90fb in PyObject_GetAttrString (v=0x7fffb186b790, name=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/object.c:846
#3  0x00005555555e5beb in PyObject_HasAttrString (v=<optimized out>, name=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/object.c:854
#4  0x00007fffdc92377b in hook_name(_object*) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#5  0x00007fffdc923dcc in torch::autograd::PyFunctionPreHook::operator()(std::vector<at::Tensor, std::allocator<at::Tensor> > const&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#6  0x00007fffd60cb1ff in torch::autograd::AccumulateGrad::apply(std::vector<at::Tensor, std::allocator<at::Tensor> >&&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#7  0x00007fffd60c651b in torch::autograd::Node::operator()(std::vector<at::Tensor, std::allocator<at::Tensor> >&&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#8  0x00007fffd60c1dc3 in torch::autograd::Engine::evaluate_function(std::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&, std::shared_ptr<torch::autograd::ReadyQueue> const&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#9  0x00007fffd60c2698 in torch::autograd::Engine::thread_main(std::shared_ptr<torch::autograd::GraphTask> const&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#10 0x00007fffd60bd481 in torch::autograd::Engine::execute_with_graph_task(std::shared_ptr<torch::autograd::GraphTask> const&, std::shared_ptr<torch::autograd::Node>, torch::autograd::InputBuffer&&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#11 0x00007fffdc9177c1 in torch::autograd::python::PythonEngine::execute_with_graph_task(std::shared_ptr<torch::autograd::GraphTask> const&, std::shared_ptr<torch::autograd::Node>, torch::autograd::InputBuffer&&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#12 0x00007fffd60bf375 in torch::autograd::Engine::execute(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, bool, bool, bool, std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#13 0x00007fffdc917711 in torch::autograd::python::PythonEngine::execute(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, bool, bool, bool, std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#14 0x00007fffdc91818c in THPEngine_run_backward(_object*, _object*, _object*) () from /home/nelhage/sw/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#15 0x00005555556d3f76 in cfunction_call_varargs (kwargs=<optimized out>, args=<optimized out>, func=0x7ffff74b58b0) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:742
#16 PyCFunction_Call (func=0x7ffff74b58b0, args=<optimized out>, kwargs=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:772
#17 0x000055555569185f in _PyObject_MakeTpCall (callable=0x7ffff74b58b0, args=<optimized out>, nargs=<optimized out>, keywords=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:159
#18 0x0000555555715b90 in _PyObject_Vectorcall (kwnames=0x7fffb6e94400, nargsf=<optimized out>, args=<optimized out>, callable=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Include/cpython/abstract.h:125
#19 call_function (kwnames=0x7fffb6e94400, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4963
#20 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:3515
#21 0x00005555556dea92 in _PyEval_EvalCodeWithName (_co=0x7fffb2384660, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kwnames=0x7fffb6e2de98, kwargs=0x7ffff75807c8, kwcount=<optimized out>, kwstep=1, defs=0x7fffb23883c8, defcount=5, kwdefs=0x0, closure=0x0, name=0x7ffff74ae7f0, qualname=0x7ffff74ae7f0) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4298
#22 0x00005555556df943 in _PyFunction_Vectorcall (func=<optimized out>, stack=0x7ffff75807a8, nargsf=<optimized out>, kwnames=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:435
#23 0x000055555565411a in _PyObject_Vectorcall (kwnames=0x7fffb6e2de80, nargsf=<optimized out>, args=<optimized out>, callable=0x7fffb2355ca0) at /tmp/build/80754af9/python_1599203911753/work/Include/cpython/abstract.h:127
#24 call_function (kwnames=0x7fffb6e2de80, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4963
#25 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:3515
#26 0x00005555556dea92 in _PyEval_EvalCodeWithName (_co=0x7fffb6e35450, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kwnames=0x0, kwargs=0x7ffff752e3c0, kwcount=<optimized out>, kwstep=1, defs=0x7fffb6e39aa8, defcount=4, kwdefs=0x0, closure=0x0, name=0x7ffff74ae7f0, qualname=0x7fffb6f83970) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4298
#27 0x00005555556df943 in _PyFunction_Vectorcall (func=<optimized out>, stack=0x7ffff752e3b8, nargsf=<optimized out>, kwnames=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:435
#28 0x000055555565475e in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7ffff752e3b8, callable=0x7fffb6d64790) at /tmp/build/80754af9/python_1599203911753/work/Include/cpython/abstract.h:127
#29 call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=0x5555558ea520) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4963
#30 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:3486
#31 0x00005555556df86b in function_code_fastcall (globals=<optimized out>, nargs=0, args=<optimized out>, co=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:283
#32 _PyFunction_Vectorcall (func=<optimized out>, stack=0x7ffff75bc5b0, nargsf=<optimized out>, kwnames=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:410
#33 0x0000555555652b84 in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7ffff75bc5b0, callable=0x7fffb18c3c10) at /tmp/build/80754af9/python_1599203911753/work/Include/cpython/abstract.h:127
#34 call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=0x5555558ea520) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4963
#35 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:3500
#36 0x00005555556dea92 in _PyEval_EvalCodeWithName (_co=0x7ffff74943a0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kwnames=0x0, kwargs=0x0, kwcount=<optimized out>, kwstep=2, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=0x0, qualname=0x0) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4298
#37 0x00005555556df754 in PyEval_EvalCodeEx (_co=<optimized out>, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kws=<optimized out>, kwcount=0, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4327
#38 0x000055555576dedc in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:718
#39 0x000055555576df84 in run_eval_code_obj (co=0x7ffff74943a0, globals=0x7ffff761f2c0, locals=0x7ffff761f2c0) at /tmp/build/80754af9/python_1599203911753/work/Python/pythonrun.c:1125
#40 0x00005555557a01f4 in run_mod (mod=<optimized out>, filename=<optimized out>, globals=0x7ffff761f2c0, locals=0x7ffff761f2c0, flags=<optimized out>, arena=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Python/pythonrun.c:1147
#41 0x00005555556686e1 in PyRun_FileExFlags (fp=0x5555558eaf90, filename_str=<optimized out>, start=<optimized out>, globals=0x7ffff761f2c0, locals=0x7ffff761f2c0, closeit=1, flags=0x7fffffffdbd8) at /tmp/build/80754af9/python_1599203911753/work/Python/pythonrun.c:1063
#42 0x0000555555668ac6 in PyRun_SimpleFileExFlags (fp=0x5555558eaf90, filename=<optimized out>, closeit=1, flags=0x7fffffffdbd8) at /tmp/build/80754af9/python_1599203911753/work/Python/pythonrun.c:428
#43 0x000055555566998b in pymain_run_file (cf=0x7fffffffdbd8, config=0x5555558eb5b0) at /tmp/build/80754af9/python_1599203911753/work/Modules/main.c:387
#44 pymain_run_python (exitcode=0x7fffffffdbd0) at /tmp/build/80754af9/python_1599203911753/work/Modules/main.c:612
#45 Py_RunMain () at /tmp/build/80754af9/python_1599203911753/work/Modules/main.c:691
#46 0x00005555557a2d19 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Modules/main.c:1137
#47 0x00007ffff7db00b3 in __libc_start_main (main=0x55555566a460 <main>, argc=2, argv=0x7fffffffddd8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffddc8) at ../csu/libc-start.c:308
#48 0x0000555555732e93 in _start () at ../sysdeps/x86_64/elf/start.S:103
```

## Expected behavior

It doesn't segfault. If a hook removing itself is considered an error, it should raise an exception.

## Environment

```
PyTorch version: 1.9.0a0+git6af4e99
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 13.0.0-++20210418052640+d480f968ad8b-1~exp1~20210418153358.383
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080 SUPER
Nvidia driver version: 450.119.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.9.0a0+git9def776
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.2                      256  
[conda] mkl-include               2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     1.9.0a0+git9def776          pypi_0    pypi
```
## Additional context

Tested on PyTorch built from master  as of 9def776cd6d775779db9839d942a01b4b082175c as well as on 1.8.1

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano

**Comment by albanD** [comment]
Hi,

Thanks for the report.
It is indeed worrying and most likely a bug in the way we iterate the hooks. We definitely want to fix that.
